app/src/main/java/com/example/mvpmusicapp/ui/MainActivity.kt

```diff
@@ -0,0 +1,248 @@
+                    storageRequest
+                )
+            }
+        }
+        bindService(musicIntent, serviceConnection, Context.BIND_AUTO_CREATE)
+        registerReceiver(musicReceiver, IntentFilter(getString(R.string.intent_action)))
+    }
+
+    private fun setConstraintMusicDialog() {
+        binding.apply {
+            cardPlayMusic.visibility = View.VISIBLE
+            imagePlayMusic.setImageResource(R.drawable.ic_pause)
+            updateMusicName()
+        }
+
+        val constraintSet = ConstraintSet()
+        constraintSet.clone(binding.constraintMainParent)
+        constraintSet.connect(
+            R.id.recycler_song,
+            ConstraintSet.BOTTOM,
+            R.id.card_play_music,
+            ConstraintSet.TOP
+        )
+        constraintSet.applyTo(binding.constraintMainParent)
+    }
+
+    override fun getAllSongList(songList: List<Song>) {
+        this.songList = songList.toMutableList()
+        songAdapter!!.setSongList(this.songList)
+        binding.recyclerSong.adapter = songAdapter
+        musicService?.setSongList(this.songList)
+    }
+
+    override fun showErrorMessage(message: String) {
+        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
+    }
+
+    override fun setPlayButton() {
+        musicService?.pauseMusic()
+        binding.imagePlayMusic.setImageResource(R.drawable.ic_play_button)
+    }
+
+    override fun setPauseButton() {
+        musicService?.resumeMusic()
+        binding.imagePlayMusic.setImageResource(R.drawable.ic_pause)
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        if (requestCode == storageRequest && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            songPresenter!!.getLocalSong()
+        }
+    }
+
+    private fun updateMusicName() {
+        binding.apply {
+            textMusicName.text = songList[currentSong].name
+            textMusicInfor.text = resources.getString(
+                R.string.text_music_infor,
+                convertToDurationFormat(
+                    this@MainActivity,
+                    musicService?.getDuration()!!.toDouble()
+                ),
+                songList[currentSong].artist
+            )
+        }
+    }
+
+    private fun updateProgress() {
+        val musicTime = Timer()
+        musicTime.scheduleAtFixedRate(object : TimerTask() {
+            override fun run() {
+                runOnUiThread {
+                    if (musicService?.currentPosition()!! < musicService?.getDuration()!!) {
+                        binding.progressMusic.progress = musicService?.currentPosition()!!
+                    } else {
+                        binding.progressMusic.progress = 0
+                        musicTime.cancel()
+                        musicTime.purge()
+                    }
+                }
+            }
+
+        }, 0, 1000)
```

Comment:
hard code

Reply:
cái này t thử khai báo constant thì nó báo bad timer

Reply:
```suggestion
const val TIMER_PERIOD = 1000L
```
khai báo luôn trong file MainActivity